drift-codec/src/main/java/io/airlift/drift/codec/internal/compiler/ThriftCodecByteCodeGenerator.java

```diff
@@ -240,169 +252,147 @@ private FieldDefinition declareTypeField()
-                        .loadThis()
-                        .getField(codecType, typeField)
-                        .retObject());
+        MethodDefinition method = classDefinition.declareMethod(a(PUBLIC), "getType", type(ThriftType.class));
+        method.getBody().append(method.getThis()
+                .getField(typeField)
+                .ret());
     }
 
     /**
      * Defines the read method for a struct.
      */
     private void defineReadStructMethod()
     {
-        MethodDefinition read = new MethodDefinition(
-                a(PUBLIC),
-                "read",
-                structType,
-                arg("protocol", TProtocolReader.class)
-        ).addException(Exception.class);
-
-        // TProtocolReader reader = new TProtocolReader(protocol);
-        read.addLocalVariable(type(ProtocolReader.class), "reader");
-        read.newObject(ProtocolReader.class);
-        read.dup();
-        read.loadVariable("protocol");
-        read.invokeConstructor(type(ProtocolReader.class), type(TProtocolReader.class));
-        read.storeVariable("reader");
+        Parameter protocol = arg("protocol", TProtocolReader.class);
+
+        MethodDefinition method = classDefinition.declareMethod(a(PUBLIC), "read", structType, protocol)
+                .addException(Exception.class);
+
+        // ProtocolReader reader = new ProtocolReader(protocol);
+        Variable reader = method.getScope().declareVariable(ProtocolReader.class, "reader");
+        method.getBody().append(reader.set(newInstance(ProtocolReader.class, protocol)));
 
         // read all of the data in to local variables
-        Map<Short, LocalVariableDefinition> structData = readFieldValues(read);
+        Map<Short, Variable> structData = readFieldValues(method, reader);
 
         // build the struct
-        LocalVariableDefinition result = buildStruct(read, structData);
+        Variable result = buildStruct(method, structData);
 
-        // push instance on stack, and return it
-        read.loadVariable(result).retObject();
-
-        classDefinition.addMethod(read);
+        // return the instance
+        method.getBody().append(result.ret());
     }
 
     /**
      * Defines the code to read all of the data from the protocol into local variables.
      */
-    private Map<Short, LocalVariableDefinition> readFieldValues(MethodDefinition read)
+    private Map<Short, Variable> readFieldValues(MethodDefinition method, Variable reader)
     {
-        LocalVariableDefinition protocol = read.getLocalVariable("reader");
-
         // declare and init local variables here
-        Map<Short, LocalVariableDefinition> structData = new TreeMap<>();
+        Map<Short, Variable> structData = new TreeMap<>();
         for (ThriftFieldMetadata field : metadata.getFields(THRIFT_FIELD)) {
-            LocalVariableDefinition variable = read.addInitializedLocalVariable(
+            Variable variable = method.getScope().declareVariable(
```

Comment:
There is another form of this method with an initial value: `declareVariable(String variableName, BytecodeBlock block, BytecodeExpression initialValue)`.. not sure this would work or be better, just wanted to make sure you know.